### PR TITLE
Display Cropping & Hot Key Core Options + related fixes

### PR DESCRIPTION
**Pull request by HVR88**
Full details of changes can be read here: https://github.com/HVR88/HVR88_vice-libretro_extras/blob/master/readme.md

**In summary:**

Goal: To provide options to fit/fill any monitor of any aspect ratio, size or resolution, with perfectly scaled, non-distorted emulator screens at each platform’s correct aspect ratio - the resulting output should closely match the proportions/dimensions originally seen on a CRT/TV.

## Additions & Changes to Codebase ##

* UI & Cropping code added in **libretro-core.c**
* Hot Key toggle & cycling code in **/vice/src/arch/libretro/retrostubs.c**
* UI: Moved Aspect Ratio setting above “Display Border” for a more hierarchical order for the Video Options
* UI: add description to Aspect Ratio
* UI: changed description for Display Borders to warn of performance issues with this feature & recommend Border crop features instead
* UI: Renamed Zoom Mode to “Crop Vertical Borders” for accuracy and to fit with the new border cropping features (internal variable/option names unchanged for compatibility with existing opt files)
* UI: added “Crop Horizontal Borders” above the vertical setting
* UI: The option labels have been renamed to make them clearer: OFF, Small Crop, Medium Crop, Full Crop
* UI: Added “Manual Crop” option value (# 4) to “Crop Vertical Borders” which will take values from two following core options
* UI: added “Manual Top Cropping” option - values in pixels 0-60 - NTSC and PAL “Full” border sizes specially marked
* UI: added “Manual Bottom Cropping”  option - values in pixels 0-60 - NTSC and PAL “Full” border sizes specially marked
* UI: added 2 new hot keys, one for Horizontal crop mode Cycling and one for Vertical crop mode (zoom mode) cycling
* UI: Relabeled Zoom Mode toggle hot key to include “(Horizontal + Vertical)” note
* Old “Zoom Mode” code converted to Vertical Cropping option
* Added new Horizontal conditional options
* Vertical cropping conditions affect only vertical dimensions & offsets - tailored/tested presets for each platform
* Horizontal cropping conditions affect only horizontal dimensions & offsets - tailored/tested presets for each platform
* Manual Top and Bottom Cropping calculated & available for C64/128 platforms only (can easily be extended to other platforms if needed)
* Shifted the mapper numerical array indexes for the datasette hot keys and increase the mapper array size by 2
* Changed Zoom Mode toggle in retrostubs to remove redundant conditions & include horizontal mode so it toggles H+V at the same time - now it’s possible to do a 2-dimensional zoom on any display
* Retrostubs includes the new cycling code and additions for the cycle hot keys as necessary
* Fixed existing crop dimensions & offsets - borders pixel-exact now when cropped for all platforms
  * Pre-set crop dimensions for c64/128 selected & tested against large library of game content & for acceptable fit into standard monitor dimensions/orientations
  * Pre-set crop dimensions for xplus4 selected to match results of c64
  * Pre-set crop dimensions for Vic-20 selected based on previous values from Zoom Mode
* Fixed: C64MODEL_C64_JAP not marked as RETRO_REGION_NTSC - missing from retro_get_region():

### CODE REVIEW

**Comment by rsn8887** [comment]
Please squash away all the commit "noise" to make this easier to understand. There shouldn't be dozens of "update readme-fork" commits update readme commits etc which do nothing in the end.

Thanks for the PR. It is great you are looking at improving the zoom/crop options.

My two cents: However, I feel this makes it a bit too complicated. We are aiming to make the core easy to use. To accomplish your goal, can you not simply add a single core option "Crop Horizontal Borders" in addition to the existing options? Instead of adding three and having to choose a crop mode separately from the other options, the user could just select the zoom mode which is basically vertical cropping, and then also select horizontal cropping.

If you really want the pixel selections, maybe just add them to the existing zoom mode choices and new horizontal crop choices add the end "OFF, small, medium, large, 1px, 2px... etc."

I am not sure we even want to let users select exact pixel counts, it seems to be too cumbersome. Instead a few sensible defaults like "small, medium, large" might be enough for the new horizontal crop option. The pixel-exact options are already in Retroarch's video and custom aspect functionalit, and also require single pixel settings there, which is just too much hassle for anyone IMO. 

@sonninnos what do you think?

File: libretro/libretro-core.c [hunk 1]
```diff
@@ -1396,19 +1412,40 @@ void retro_set_environment(retro_environment_t cb)
...
+         "auto"
+      },
+#endif
+
+#if !defined(__PET__) && !defined(__CBM2__)     // This should always be left ON, serious performance and graphics glitches when OFF
       {
          "vice_border",
          "Display Borders",
-         "All cores except 'x64sc' will reset when this option is changed.",
+         "Leave 'ON' otherwise may affect performance - Use Border Cropping options instead\n"
```
Comment (rsn8887) [inline libretro/libretro-core.c:1447]: Please revert this change. This does not affect performance. Borders off is a standard vice option, no need to tell people to leave it "ON"

**Comment by HVR88** [comment]
The commits on the readme shouldn't be in here. That was a by-product of editing live on the web site earlier.

The RA options for video scaling are, to put it bluntly, useless. They should absolutely never be touched unless the core is deficient. These changes make the core full featured with respect to the frame buffer.

The pixels options are added as an extension of the VERTICAL crop only for very specific reason. This allows Cycle hot-keys to move between the different modes constrained to a sensible total, 4. Having pixel amounts in there wouldn't allow being able to cycle all the modes - sure, it could be hard-limited to only doing the preset modes, but it would be ugly.

I've been developing software for 30+ years and simplicity and customer support is of paramount importance. When you handle replying to customers directly you get an appreciation for simplicity. But these options are hidden by default, so that they're already advanced in nature. They also offer the ability to eliminate the buggy "Display Borders" option.

I could make the pixel options hidden until the "Manual" vertical preset is selected, but because the core options menu doesn't have live-updating like RA's own menus, I decided not to do that (same as all the other options in there that expose more features.)

File: libretro/libretro-core.c [hunk 2]
```diff
@@ -1417,34 +1454,187 @@ void retro_set_environment(retro_environment_t cb)
...
+  #endif
             { NULL, NULL },
          },
          "none"
       },
+
+  #if defined(__X64__) || defined(__X64SC__) || defined(__X128__)
+      {
+         "vice_bordercrop_top",         // Manual Pixel settings for Top Cropping option
+         "Manual Top Cropping",
```
Comment (rsn8887) [inline libretro/libretro-core.c:1494]: Why is pixel selection needed as a separate option? Is small, medium, full not enough? If you really need pixel values, maybe add them to the other choices of "small, medium, full" e.g. small, medium, full, 0 px, 1px, so the pixel values just appear as additional choices to the crop option.

File: libretro/libretro-core.c [hunk 3]
```diff
@@ -1417,34 +1454,187 @@ void retro_set_environment(retro_environment_t cb)
...
+            { "59", "59 px" },
+            { "60", "60 px" },      // same as variable max_crop_pixels
+            { NULL, NULL },
+         },
+         "none"
+      },
+
+      {
+         "vice_bordercrop_bottom",          // Manual Pixel settings for Bottom Cropping option
+         "Manual Bottom Cropping",
```
Comment (rsn8887) [inline libretro/libretro-core.c:1565]: This is too complicated to use IMO. Just provide a few default sensible crop options. Not even standalone Vice has this granularity. Why would it be necessary?

File: libretro/libretro-core.c [hunk 4]
```diff
@@ -1968,15 +2158,36 @@ void retro_set_environment(retro_environment_t cb)
...
-         "Hotkey: Toggle Zoom Mode",
-         "Press the mapped key to toggle zoom mode.",
+         "Hotkey: Toggle Zoom Mode (Horizontal + Vertical Cropping)",
+         "Press the mapped key to toggle Zoom ON/OFF.",
          {{ NULL, NULL }},
          "---"
       },
+
+      {
+         "vice_mapper_bcrop_horiz_mode_cycle",
```
Comment (rsn8887) [inline libretro/libretro-core.c:2173]: Does this hotkey choice actually change the option in quickmenu->options? For example, if the user uses the hotkey, then opens and closes quickmenu, does that reset the option or does it actually stay?

File: libretro/libretro-core.c [hunk 5]
```diff
@@ -4005,11 +4282,7 @@ static struct retro_disk_control_ext_callback diskControlExt = {
 
 static void fallback_log(enum retro_log_level level, const char *fmt, ...)
 {
-   (void)level;
-   va_list va;
-   va_start(va, fmt);
-   vfprintf(stderr, fmt, va);
-   va_end(va);
+   /* stub */
```
Comment (rsn8887) [inline libretro/libretro-core.c:4285]: Revert. Why are you removing this functionality?

File: libretro/libretro-core.c (hunk 2 above)
Comment (rsn8887) [reply libretro/libretro-core.c:1494]: Also maybe just give a few pixel values not every. E.g. 2, 4, 6, 8, etc.

**Comment by rsn8887** [comment]
Can you check if maybe #230 zoom refactor already fixes your needs with horizontal cropping?

File: libretro/libretro-core.c (hunk 1 above)
Comment (HVR88) [reply libretro/libretro-core.c:1447]: If you do the testing, you'll see that function is buggy. It absolutely reduces performance, including sprite glitching.

@sonninnos proposed eliminating that option. I only Clarified that it should always be left on because of its adverse effects on performance and it being ON a requirement to use any zoom/cropping.

File: libretro/libretro-core.c (hunk 2 above)
Comment (HVR88) [reply libretro/libretro-core.c:1494]: I did have it every 2 in development - but RA's menu selection is so fast using the new UI, that there's nothing inherently wrong with individual pixels.
The precision is necessary to accommodate overrides for specific content.  These settings have now been tested with hundreds of games.

File: libretro/libretro-core.c (hunk 3 above)
Comment (HVR88) [reply libretro/libretro-core.c:1565]: Why would you target the inabilities of stand-alone vice?

File: libretro/libretro-core.c (hunk 4 above)
Comment (HVR88) [reply libretro/libretro-core.c:2173]: By design, like the original Zoom Hot Key Tony implemented, it does NOT alter the menu setting. That would not be the correct thing to do as Core Options are saved by RA and pressing some hot-keys during a game shouldn't undo what you've already previously saved.

File: libretro/libretro-core.c (hunk 1 above)
Comment (rsn8887) [reply libretro/libretro-core.c:1447]: Ok I thought it was fine unless the old X64 was used (not X64sc)

File: libretro/libretro-core.c (hunk 5 above)
Comment (HVR88) [reply libretro/libretro-core.c:4285]: I didn't touch that code. Is that part of an upstream sync?

File: libretro/libretro-core.c (hunk 2 above)
Comment (rsn8887) [reply libretro/libretro-core.c:1494]: Ok I still think these should be added to existing zoom options just as extra selections to keep it simple.

Comment (HVR88) [reply libretro/libretro-core.c:1494]: Please don't evaluate these changes without first having tried the code live. Check-out my fork compile it and try it. There's nothing complicated about the presentation nor use. It was created and refined and tested to be simple to understand, concise and quick to change.

**Comment by sonninnos** [comment]
1. Zoom used to be in cycle mode, but was changed to toggle, because cycling is more annoying to use in reality
2. Old opt-compatibility is broken when "maximum" is changed to "full"
3. I don't really enjoy the renaming of zoom, because then I'd have to change it in PUAE too
4. I don't see the real need of pixel fidelity cropping

Other than that, why not, I guess.

**Comment by HVR88** [comment]
> Can you check if maybe #230 zoom refactor already fixes your needs with horizontal cropping?

Tony's addition of a simple Horizontal does not fix the need for what I've added.  I proposed what Tony added last week and because I considered it insufficient I asked that the change not be considered and that I would take the task of implementing a robust solution myself.

**Comment by HVR88** [comment]
> 1. Zoom used to be in cycle mode, but was changed to toggle, because cycling is more annoying to use in reality

Toggle remains, which is great, because it works to snap everything OFF and back to what's set in the options.

Cycle is for each dimension  and works very well.

> 2. Old opt-compatibility is broken when "maximum" is changed to "full"

This was an oversight in the opt value. It should have been left as "maximium" even though the UI label is changed.  Easy to correct.

> 3. I don't really enjoy the renaming of zoom, because then I'd have to change it in PUAE too

Zoom didn't make sense before because it wasn't zoom at all.  It's still not, so I've tried to avoid using it.

> 4. I don't see the real need of pixel fidelity cropping

Again, you have to be familiar with the affected content available out there. I came into this after long discussions covering hundreds of problem titles that had no way to be presented properly using vice-libretro and RA.  Check the link at the bottom of my readme for one public discussion.  RA's own scaling options are poorly implemented and ill-conceived. This obviates the need to touch anything in RA and instead leave it set to Core Provided.

File: libretro/libretro-core.c (hunk 1 above)
Comment (HVR88) [reply libretro/libretro-core.c:1447]: It doesn't matter which C64 core is used AFAIK.  Simple test, try Commando in ViceGTK or RA.

I prefer Tony's change for this however, which is to hard-set it ON and remove it from the menu.

File: libretro/libretro-core.c (hunk 2 above)
Comment (HVR88) [reply libretro/libretro-core.c:1494]: Another reason they're separate is to not complicate the Vertical options list. It's nice and simple at 4 presets.  If you only want to use presets, then you can do so and not be bothered to look at any pixels.  These menus are closed and start at 0. You have to purposefully open them or change the values to use.  No complication needed.

**Comment by HVR88** [comment]
I just took a look at the changes from #230 and my branch is a lot cleaner and doesn't have special-cases for 16:9 or other DAR because they all just work.  IMO, that's really not the way to address this.

**Comment by rsn8887** [comment]
Instead of pixel increments, can an option "Auto" be implemented for cropping? The auto option would just crop whatever is black in the border. This exists in PUAE, but of course has to be done with some thought about performance (e.g. don't want to check every single pixel every frame if black or not).

**Comment by Zombeaver** [comment]
This was prompted largely by [this project](https://forums.launchbox-app.com/topic/49324-c64-dreams-massive-curated-c64-collection/) which has been ongoing for several years now and will likely continue for years to come.

I can tell you that I've spent quite a bit of time testing @HVR88's changes and they are absolutely an improvement as far as I'm concerned and for the purposes of that project. They're simultaneously easy to use and grant a major amount of control that I can tell you, as someone that has setup literally thousands of games, _is necessary_. The three pre-defined levels are not sufficient. The way devs made use of the space is really all over the place, so allowing this amount of granularity to compensate is a _massive_ boon. Adjusting the size at Retroarch's video output level is not a good option either because it instantly becomes resolution-dependent at that point. With these changes, the video can be cropped to the content for everyone, across the board, regardless of their native resolution or aspect ratio. You leave the aspect ratio on core provided and you're done - if you want additional granular control it's at your disposal or the existing levels are intact if not.

> Sure you also tested how the VKBD and the statusbar look and work with the manual croppings?

VKBD still works, although it'll be cropped too if you keep cutting beyond a certain point - 41 top / 43 bottom to be specific. That's beyond "full/maximum". Statusbar works.

**Comment by rsn8887** [comment]
Vkbd should never be able to be cropped.

**Comment by Zombeaver** [comment]
> It should not be possible for the user to crop Vkbd beyond usability.

You mean exactly like what happens if you were to adjust your video settings with custom values? Because it gets cropped in that scenario as well - even worse in that case because the VKBD positioning is still based on the content which is unaltered.

I would agree, however, that if there's a way to make it scale even after adjustments, that _would_ be preferable.

**Comment by HVR88** [comment]
> It should not be possible for the user to crop Vkbd beyond usability.

The vkbd should be rendered AFTER the native frame buffer.  The cropping is affecting original platform's resolution/frame. The VKBD is a construct for the running system and shouldn't be affected by the cropping.  It sound like maybe that needs to be refactored.

**Comment by HVR88** [comment]
> Instead of pixel increments, can an option "Auto" be implemented for cropping? The auto option would just crop whatever is black in the border. This exists in PUAE, but of course has to be done with some thought about performance (e.g. don't want to check every single pixel every frame if black or not). Just a thought.

A lot more would also be possible if the core were able to check the current desktop resolution.  An active/polling solution is, IMO far too heavy handed an approach.

**Comment by rsn8887** [comment]
The core has no concept of host resolution. So of course the vkbd, being implemented in the core, must be rendered in core frame buffer, the only frame buffer there is.

**Comment by sonninnos** [comment]
If pixel precision is so important, why only top & bottom? Why not also left & right? Why the need of cropping more than the actual VIC-II/VIC/TED borders? And if they are adjusted per-game basis, what is the need for hotkeys then?

I'm ok with having separate crop options, but they don't have to have anything to do with the current zoom, which is still zoom.

**Comment by rsn8887** [comment]
> The commits on the readme shouldn't be in here. That was a by-product of editing live on the web site earlier. Happy to try and eliminate them if you can offer a suggestion of how to do so. That said, they don't affect the ability to browse/evaluate the latest condition of the two changed files in this request. **I have tried using GitKraken to squash but the option won't appear in the context menu for any commits.** When I issued the pull I didn't expect to see my fork-specifc commits to come up.

`git rebase -i HEAD~XX`

where XX is how large commit history you want to edit (HEAD~4 means you will see the last 4 commits).

then follow instructions in the editor that pops up

use d to drop commits, s to squash etc.
save and exit editor
`git push -f origin master` to force-push changes to your branch

> 
> The RA options for video scaling are, to put it bluntly, useless. They should absolutely never be touched unless the core is deficient. These changes make the core full featured with respect to the frame buffer. There is no reason whatsoever to have RA's video/scaling set to anything other than "Core Provided" using my changes.

Agreed.

> 
> The pixels options are added as an extension of the VERTICAL crop only for very specific reason. This allows Cycle hot-keys to move between the different modes constrained to a sensible total, 4. Having pixel amounts in there wouldn't allow being able to cycle all the modes - sure, it could be hard-limited to only doing the preset modes, but it would be ugly.
> 

Makes sense.

> I've been developing software for 30+ years and simplicity and customer support is of paramount importance. When you handle replying to customers directly you get an appreciation for simplicity. But these options are hidden by default, so that they're already advanced in nature. They also offer the ability to eliminate the buggy "Display Borders" option.

Ok.

> 
> I could make the pixel options hidden until the "Manual" vertical preset is selected, but because the core options menu doesn't have live-updating like RA's own menus, I decided not to do that (same as all the other options in there that expose more features.)

I think making them hidden by default would be good, like an `Show advanced cropping options` ON/OFF switch. Have to exit and re-enter menu, but I think that is always the case in Retroarch and many core options do that.

**Comment by Zombeaver** [comment]
> If pixel precision is so important, why only top & bottom? Why not also left & right? Why the need of cropping more than the actual VIC-II/VIC/TED borders?

Technically you could. We talked about that while he was working on it. It wasn't included because there are very, very few scenarios where it would really be warranted. _Far_ fewer than top and bottom. There's literally one game out of the 2000 in C64 Dreams that would warrant it - Jar's Revenge - and the existing pre-defined levels are sufficient in that scenario. It's the only game out of all of them that actually uses extra horizontal border.

I wouldn't be _opposed_ to it, but the use scenario is extremely small, much more so than top and bottom.

**Comment by HVR88** [comment]
> The core has no concept of host resolution. So of course the vkbd, being implemented in the core, must be rendered in core frame buffer, the only frame buffer there is.

You should consider making the vkbd dynamic. When the screen is cropped, the keyboard actually becomes more usable on a touch device because they virtual keys are rendered larger. It is only by the xtreme overcropping that the keyboard is encroached.

However, if the keyboard is resized to be proportional to the frame buffer, even with extreme cropping, it could be made close to its original size (when borders are showing) and nothing would be lost.

**Comment by HVR88** [comment]
> > If pixel precision is so important, why only top & bottom? Why not also left & right? Why the need of cropping more than the actual VIC-II/VIC/TED borders?
> 
> Technically you could. We talked about that while he was working on it. It wasn't included because there are very, very few scenarios where it would really be warranted. 

Frogger Arcade could also be fixed and centered with manual horizontal cropping too. :)  But ya, 2 games doesn't make it worth it.  

We know that Top and Bottom are affected by the majority of games in some way.

**Comment by rsn8887** [comment]
Ok let's aim at making all this advanced cropping pixel selection etc a bunch of additional options that appear when a "Show Advanced Cropping Options" core options is set to ON.

In there can be options to select top and bottom pixels etc. and each of those options has a selection "Default" which means it behaves as before this PR, and the pixel values and other options to override cropping with the advanced functionality.

Does that sound reasonable?

**Comment by Zombeaver** [comment]
I think making it accessible in an advanced section is perfectly reasonable, so long as the functionality is still there that seems fine.

> Frogger Arcade could also be fixed and centered with manual horizontal cropping too. :) But ya, 2 games doesn't make it worth it.

That's true, that's another use case, although that one's a little different since it doesn't use extra horizontal space it just... uses the normal space oddly. I think it would be overkill in general though.

**Comment by HVR88** [comment]
> Ok let's aim at making all this advanced cropping pixel selection etc a bunch of additional options that appear when a "Show Advanced Cropping Options" core options is set to ON.

Instead of another option to show it, it would be easy to make that dependent on th Vertical preset being set to "Manual" - accomplishes the same thing without yet another option to b stuck on the screen showing only a binary setting. "Manual" could potentially be called "Advanced" but that's somewhat generic and doesn't convey what's really happening (or rather going to happen).

It will have to include a note about refreshing to show the options.  Will core options ever get live-updating menus in RA?

I can also definitely set defaults instead of starting at 0.

> Then there's room to also add manual horizontal cropping pixel values if you want in there. Because it is clear that stuff only appears when the user really wants it, so it won't be confusing.

I'll take a look at this one. Easy to implement, just not sure it's terribly useful without some insane numerical range, because as mentioned, where it makes sense is for something like Frogger which is a vertically-oriented release. But I honestly don't know of any other.  Zomb is intimately familiar with more titles as until recently I'd been away from the C64 since 1989.

**Comment by rsn8887** [comment]
Yes all changes from Tony should be there, basically, can you rebase on current master and fix the conflicts?

I like the options that sonninnos added, the new  Zoom crop: "horizontal, vertical, both..." should stay. The zoom option should stay as-is. All that is the simple use case for users who don't care about pixel-level cropping.

Then we add an option: Show advanced cropping options YES/NO

If enabled, there are options: Manual Zoom Crop Top, Manual Zoom Crop Bottom, and if you want Manual Zoom Crop Left and Manual Zoom Crop Right options with choices "Default, 1px, 2px" etc.. If the user selects any pixel choices here, that overrides the default cropping behavior and uses the pixel values instead.

I am open to suggestions to make that better, but that is what I think now.

**Comment by rsn8887** [comment]
Never mind the Show Advanced Cropping Option thing is a bad idea. These options are already behind the "Show Video Options" guard. We don't want two of those nested.

Ok maybe just add a "manual" option to the new Zoom Crop option? And then always display the options underneath "Manual Zoom Crop Top" etc. with the pixel values.

**Comment by HVR88** [comment]
I thought I'd mentioned already that there was already a VIdeo Options hiding property.  This is another reason why I didn't hide the manual pixel selections - it's a lot of jumping in and out of the core just two hide two lines.

> Yes all changes from Tony should be there, basically, can you rebase on current master and fix the conflicts?

I'll have to think about this tomorrow.  I don't think it'll be easy nor clean to present these options along the latest made by Tony. First there's the way the code has changed in his latest commit - it's a little bit of a run-around, IMO.

I'd have to make an advanced switch circumvent all his new code and hide the zoom options. Then I'd have to include what I have now.  It's doable, but IMO, the code is going to get a lot more complex in comparison to little to no benefit in the UI/UX. But to keep what was added in #230 this would have to be done. Those options aren't compatible with my cropping presets and manual cropping and can't easily be made to be.

I believe the cleanest approach is what I had suggested prior.

**Comment by rsn8887** [comment]
My goal is to have the existing options as the simple use case, they already work together for somebody who just wants some zoom presets that work with different display aspects.

Then your option is an advanced one that allows pixel values to be selected. That should be separate IMO.

The only remaining question is where to put that advanced functionality. I think it best fits in the new "Zoom Crop" option as a choice "manual crop", because that already has crop in the name.

And then always display the pixel choices underneath. But explain in the tooltip: "pixel value to use when Zoom Crop is set to manual" or so. That will be best I think.

That way we get the benefit from Sonninnos latest commits, allowing simple zoom crop that works with many displays, plus your benefit of allowing manual pixel values.

**Comment by HVR88** [comment]
I've come up with a way to do all this, but it's going to be a stupid amount of new work. It will basically involve destroying this pull request completely along with its fork. Then making a new fork from the current master and then manually adding in all my original code with a slight refactor plus an option to select between basic and advanced frame cropping. That option not only affecting what is shown in the Core Option menu, but which code path to use because they can't both exist together. 

The slight refactor to my code would be to remove all reliance on older zoom variable names (which I had left in for backwards compatibility) and replace them with new unique names. That way the code doesn't collide even when there might not be a conditional path AND opt variables can co-exist.

Doing it this way, I can make a SINGLE commit with absolutely all changes at once - I'll develop the new/updated fork locally only and test it by making local/private builds without affecting the repo. So no more commits to squash, no merging dance necessary.

To address latest points... - which don't negate what I just wrote above.

> My goal is to have the existing options as the simple use case, they already work together for somebody who just wants some zoom presets that work with different display aspects.

But they don't work now and thy didn't work before.  If you haven't tested it and if you don't have a graphics and display background, it's easy to overlook.  But tying any horizontal value to the vertical means that it can't work for any display that restricts both dimensions. A wide display will never restrict the width, so it work there. But a narrow display, such as 4:3 or common 5:4 LCD does and any amount of horizontal border means the resulting frame is not as large as it could be.

The presets I have now can fill every single pixel of a 5:4 display and also provide the largest frame possible on a 16:9 widescreen. The only way to achieve a maximal size with linked horizontal and vertical cropping is to always set the horizontal to fully remove the borders or pre-define values that work for the narrow display type - both of those solutions are inelegant and will leave someone unhappy they can't format the display to their situation. 

> Then your option is an advanced one that allows pixel values to be selected. That should be separate IMO.

It's separate now in my code.

> Sorry for the extra work fixing conflicts/incorporating your solution with those simple options from sonninnos, and rebasing, but your PR required some kind of rebase anyways, because of all those wrong commits that shouldn't be there. 

Those wrong commits had nothing to do with the code and were a reflection of a file that doesn't exist at all in the repo/pull. Easy to take out using your suggestion on the command line.

>It might appear less clean, but it will leave that simple crop option horizontal, vertical, both in there for users who don't care about the pixel values.

Which is what I implemented.  Honestly, I understand you guys have a history on this project and that's one of the reasons I was hesitant to contribute in the first place. I could easily keep my fork up to date and not share back to the main project. The features I put into this fork and the UI to accompany them are just about as simple and clean as it gets and the user experience is simple and straight forward - I even cleaned up the order and grouping of a few UI options and added additional documentation labels that, IMO, were sorely missing.

I'd go further and relabel "Aspect Ratio" to "Pixel Aspect Ratio" or "Pixel Aspect Ratio Correction" as well as remove some other code that is extraneous and incorrect (different pixel aspect ratio adjustment for border off scenario which looks disabled anyway)

If you haven't looked and compared the code to the latest pull you really should. And I don't think you're getting the full picture if you haven't tried the build live. Especially on different displays and with different content.

**Comment by sonninnos** [comment]
I could add the manual crop options too, won't be a problem.

**Comment by rsn8887** [comment]
I merged another attempt #231 by @sonninnos to implement working manual crop which implements the manual crop in a clean way with minimal code changes. I think that probably takes care of this. Thanks for the work and without this example here it probably wouldn’t have happened.

In the future please keep pull requests a bit cleaner, with just a few commits per PR and code changes that are only related to the features. Several small PRs, one per feature are preferred over one large one.

I am closing this PR due to #231 taking care of most of this and the recent conflicts. If anything is still missing please make new PRs, one per feature ideally.

It helps to rebase and clean your commit history in your branch before making a PR. It also helps if you make a local branch different from master for each of your PRs so that you can work on features separate from all the commit “noise” in master.

**Comment by HVR88** [comment]
I have a solve that will achieve all the goals presented on all sides, including keeping clean and maintainable source without display-dependent tweaks.

The goal:

Offer a "Zoom" feature that is simple and replicates the last commit from @sonninnos in the UI presentation.

Include the ability to toggle to an "advanced" mode whereby there is greater control for cropping.

I intended to do this with fewer options than Tony included himself. But I'll have to look at what was just created.

If what's here is incompatible, I'm just going to fork, do the work and won't be requesting another pull. I'll rename the project and it will be offered and used as an alternative to this one.

**Comment by HVR88** [comment]
Ya, so I've decided what to do. I'm going to copy the project without an uplink and maintain my own project.

I can't for one second believe I received this kind of negative feedback and now taking a look at what's been done in commit 231 is an awful mess.  I'm sorry if that's insulting, but the code is spaghetti, doesn't take into account differences of NTSC and PAL, includes new bugs and bugs that I had previously patched.  

The UI has been complicated to the nth degree with UX (that's user experience if you're new) that's become abysmal with respect to this feature and no one is going to understand what's going on.

Just bloody incredible.  All the best, but I'm done here.

**Comment by sonninnos** [comment]
Look, the zoom code was shit to begin with, and you did not make it any less shittier, instead just piled up on it, making it more unmaintainable and cemented.

Please realize this is still WIP, so no reason to get too hung up on something that might change soon anyway. Things get better in small steps as they go along. I'm open to new and better ideas, and I don't consider myself to be a know-it-all, but rather a student in progress. Surely you can understand me wanting to remove code that shouldn't have been there in the first place.

Yes, that is rather rude. Does that also mean you'll be reverting all my changes over the months for the lack of testing and understanding?

And if you honestly think the current spaghetti is worse than before for improving on, then I guess that decision is a win-win. Clearly you don't have the common interest at heart, if you are not willing to point out the bugs. More data, less barking, please. I'm too old for this shit.

**Comment by rsn8887** [comment]
This PR was 37 commits long and changed unrelated code and comments all over the place, that’s why it received negative feedback.
